src/setup_payload/AdditionalDataPayloadGenerator.cpp

```diff
@@ -0,0 +1,130 @@
+    mSerialNumberBuffer(serialNumberBuffer), mSerialNumberBufferSize(serialNumberBufferSize)
+{
+    mHashInput =
+        reinterpret_cast<char *>(chip::Platform::MemoryAlloc(ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + serialNumberBufferSize));
+}
+
+AdditionalDataPayloadGenerator::~AdditionalDataPayloadGenerator()
+{
+    if (mHashInput != nullptr)
+    {
+        chip::Platform::MemoryFree(mHashInput);
+        mHashInput = nullptr;
+    }
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateAdditionalDataPayload(PacketBufferHandle & bufferHandle)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    TLVWriter innerWriter;
+    char rotatingDeviceIdBuffer[ROTATING_DEVICE_ID_LENGTH + 1];
+    uint8_t rotatingDeviceIdBufferSize;
+
+    // Initialize TLVWriter
+    writer.Init(PacketBuffer::New());
+
+    err = writer.OpenContainer(AnonymousTag, kTLVType_Structure, innerWriter);
+    SuccessOrExit(err);
+
+    // Generating Device Rotating Id
+    err = generateRotatingDeviceId(rotatingDeviceIdBuffer, rotatingDeviceIdBufferSize);
+    SuccessOrExit(err);
+
+    // Adding the rotating device id to the TLV data
+    err = innerWriter.PutString(ContextTag(kRotatingDeviceIdTag), rotatingDeviceIdBuffer);
+    SuccessOrExit(err);
+
+    err = writer.CloseContainer(innerWriter);
+    SuccessOrExit(err);
+
+    err = writer.Finalize(&bufferHandle);
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR AdditionalDataPayloadGenerator::generateRotatingDeviceId(char * rotatingDeviceIdBuffer,
+                                                                    uint8_t & rotatingDeviceIdBufferSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    char rotationCounterStr[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH];
+
+    char output[ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH];
+    uint8_t hashOutputBuffer[kSHA256_Hash_Length];
+    uint8_t outputIndex     = 0;
+    uint8_t hashOutputIndex = 0;
+
+    sprintf(rotationCounterStr, "%d", mRotationCounter);
+    strcpy(mHashInput, rotationCounterStr);
+    strcat(mHashInput, mSerialNumberBuffer);
+
+    err = Hash_SHA256(reinterpret_cast<unsigned char *>(mHashInput),
+                      ROTATING_DEVICE_ID_COUNTER_STR_LENGTH + mSerialNumberBufferSize, hashOutputBuffer);
+    SuccessOrExit(err);
+
+    // Computing the Rotating Device Id
+    // RDI = Rotation_Counter + SuffixBytes(HMAC_SHA256(Serial_Number + Rotation_Counter), 16)
+    while (outputIndex < ROTATING_DEVICE_ID_COUNTER_STR_LENGTH && rotationCounterStr[outputIndex] != '\0')
+    {
+        output[outputIndex] = rotationCounterStr[outputIndex];
+        outputIndex++;
+    }
+
+    while (hashOutputIndex < ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH)
+    {
+        output[outputIndex++] =
+            static_cast<char>(hashOutputBuffer[kSHA256_Hash_Length - ROTATING_DEVICE_ID_HASH_SUFFIX_LENGTH - 1 + hashOutputIndex]);
+        hashOutputIndex++;
+    }
```

Comment:
This code is overall quite non-trivial with regards to counters usage and size assumptions. Please move to BufferWriter instead.

Reply:
used BufferWriter